### DISCUSSION TITLE
Running into issue with update libmesh script on HPC

### DISCUSSION THREAD

**Discussion by TheGreatCid** [idaholab/moose#28530] [Q&A General]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

I am trying to run `update_and_rebuild_libmesh.sh` on an HPC and I am running into the following error:
```
I for src/utilities/include/timpi/timpi_config.h.tmp defines
config.status: executing depfiles commands
config.status: error: in `/ascldap/users/detorre/raccoon/moose/libmesh/build/contrib/timpi':
config.status: error: Something went wrong bootstrapping makefile fragments
    for automatic dependency tracking.  If GNU make was not used, consider
    re-running the configure script with MAKE="gmake" (or whatever is
    necessary).  You can also try re-running configure with the
    '--disable-dependency-tracking' option to at least be able to build
    the package (albeit without support for automatic dependency tracking).
See `config.log' for more details
configure: WARNING: cache variable ac_cv_env_timpi_CPPFLAGS_value contains a newline
configure: error: ../../../contrib/timpi/configure failed for contrib/timpi
Running make -j 15...
make: *** No targets specified and no makefile found.  Stop.
```

I have tried both of the suggestions from the error message to no avail. 

This is the output of `diagnostics.sh`

```
##################################################################################################
Influential Environment Variables

CC=mpicc
CONDA_CHANNEL=https://conda.software.inl.gov/public
CONDA_EXE=/home/detorre/mambaforge3/bin/conda
CONDA_PYTHON_EXE=/home/detorre/mambaforge3/bin/python
CONDA_SHLVL=0
CURL_CA_BUNDLE=
CXX=mpicxx
F77=mpif77
F90=mpif90
FC=mpif90
http_proxy=http://user:pass@proxy.sandia.gov:80
https_proxy=http://user:pass@proxy.sandia.gov:80
LD_LIBRARY_PATH=/opt/toss/openmpi/4.1/gnu/lib:/usr/lib64/nvidia:/projects/netpub/gnu/11.3.0-toss3/lib64:/projects/netpub/gnu/11.3.0-toss3/lib::
MODULESHOME=/usr/share/lmod/lmod
MOOSE_JOBS=15
PATH=/opt/toss/openmpi/4.1/gnu/bin:/projects/cmake/toss4/3.25.2/bin:/projects/aue/hpc/builds/x86_64/rhel8/ba17d7f2/tooling/install/linux-rhel8-x86_64/gcc-10.3.0/python-3.9.16-a7hvvwm/bin:/projects/netpub/gnu/11.3.0-toss3/bin:/home/detorre/mambaforge3/condabin:/ascldap/users/detorre/mambaforge3/envs/gmsh2/bin:/usr/lpp/mmfs/bin:/opt/ibm/MCStore/bin:/opt/ibm/MCStore/scripts:/usr/lib64/ccache:/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin
REQUESTS_CA_BUNDLE=
SSL_CERT_FILE=

##################################################################################################
Compiler(s) (CC CXX FC F77 F90):

CC=/opt/toss/openmpi/4.1/gnu/bin/mpicc
CC -show:
gcc -I/opt/toss/openmpi/4.1/gnu/include -pthread -L/opt/toss/openmpi/4.1/gnu/lib -Wl,-rpath -Wl,/opt/toss/openmpi/4.1/gnu/lib -Wl,--enable-new-dtags -lmpi
CC version:	gcc (GCC) 11.3.0

CXX=/opt/toss/openmpi/4.1/gnu/bin/mpicxx
CXX -show:
g++ -I/opt/toss/openmpi/4.1/gnu/include -pthread -L/opt/toss/openmpi/4.1/gnu/lib -Wl,-rpath -Wl,/opt/toss/openmpi/4.1/gnu/lib -Wl,--enable-new-dtags -lmpi
CXX version:	g++ (GCC) 11.3.0

FC=/opt/toss/openmpi/4.1/gnu/bin/mpif90
FC -show:
gfortran -I/opt/toss/openmpi/4.1/gnu/include -pthread -I/opt/toss/openmpi/4.1/gnu/lib -L/opt/toss/openmpi/4.1/gnu/lib -Wl,-rpath -Wl,/opt/toss/openmpi/4.1/gnu/lib -Wl,--enable-new-dtags -lmpi_usempif08 -lmpi_usempi_ignore_tkr -lmpi_mpifh -lmpi
FC version:	GNU Fortran (GCC) 11.3.0

F77=/opt/toss/openmpi/4.1/gnu/bin/mpif77
F77 -show:
gfortran -I/opt/toss/openmpi/4.1/gnu/include -pthread -I/opt/toss/openmpi/4.1/gnu/lib -L/opt/toss/openmpi/4.1/gnu/lib -Wl,-rpath -Wl,/opt/toss/openmpi/4.1/gnu/lib -Wl,--enable-new-dtags -lmpi_usempif08 -lmpi_usempi_ignore_tkr -lmpi_mpifh -lmpi
F77 version:	GNU Fortran (GCC) 11.3.0

F90=/opt/toss/openmpi/4.1/gnu/bin/mpif90
F90 -show:
gfortran -I/opt/toss/openmpi/4.1/gnu/include -pthread -I/opt/toss/openmpi/4.1/gnu/lib -L/opt/toss/openmpi/4.1/gnu/lib -Wl,-rpath -Wl,/opt/toss/openmpi/4.1/gnu/lib -Wl,--enable-new-dtags -lmpi_usempif08 -lmpi_usempi_ignore_tkr -lmpi_mpifh -lmpi
F90 version:	GNU Fortran (GCC) 11.3.0

OK

##################################################################################################
Python Sanity Checks

Verify `/usr/bin/env python3 --version` (reporting as: Python 3.9.16),
matches versions for: `which python3 && which python`

OK

##################################################################################################
Python Modules (TestHarness, run-ability)

WARNING: python module: `yaml` not available

Warning Python packages may cause miscellaneous runtime issues.

Either install the above packages, or perhaps you have yet to activate your
moose environment: `conda activate moose`.

```

I see the error with `yaml` but that seems unrelated to this issue. Though I imagine I'll have to address that eventually. 

Any insight into the issue would be appreciated. Thanks!

**Comment by GiudGiud** [comment]
Could you please attach the config.log file for more details? 
It ll be in libmesh/

**Reply by TheGreatCid** [reply]
[config.log](https://github.com/user-attachments/files/16853436/config.log)

**Reply by GiudGiud** [reply]
i see this there
```
/home/detorre/mambaforge3/envs/gmsh2/lib/libmpi.so: undefined reference to `opal_hwloc201_hwloc_get_type_depth'
```
how is this being found? We need mpi to solely come a single compiler, here it seems you are using the system one in /opt/

**Reply by TheGreatCid** [reply]
[config.log](https://github.com/user-attachments/files/16854567/config.log)
Just fixed that issue, however the original problem still persists.

**Reply by milljm** [reply]
Can you provide a new diagnostic report? I see several environment variables in the `config.log` you posted, that should have been captured in the diagnostic script (CXXFLAGS, etc).

**Reply by TheGreatCid** [reply]
```
##################################################################################################
Influential Environment Variables

CC=mpicc
CONDA_CHANNEL=https://conda.software.inl.gov/public
CURL_CA_BUNDLE=
CXX=mpicxx
F77=mpif77
F90=mpif90
FC=mpif90
LD_LIBRARY_PATH=/opt/toss/openmpi/4.1/gnu/lib:/usr/lib64/nvidia:/projects/netpub/gnu/11.3.0-toss3/lib64:/projects/netpub/gnu/11.3.0-toss3/lib::
MODULESHOME=/usr/share/lmod/lmod
MOOSE_JOBS=6
PATH=/opt/toss/openmpi/4.1/gnu/bin:/projects/cmake/toss4/3.25.2/bin:/projects/aue/hpc/builds/x86_64/rhel8/ba17d7f2/tooling/install/linux-rhel8-x86_64/gcc-10.3.0/python-3.9.16-a7hvvwm/bin:/projects/netpub/gnu/11.3.0-toss3/bin:/usr/lpp/mmfs/bin:/opt/ibm/MCStore/bin:/opt/ibm/MCStore/scripts:/usr/lib64/ccache:/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin
REQUESTS_CA_BUNDLE=
SSL_CERT_FILE=

##################################################################################################
Compiler(s) (CC CXX FC F77 F90):

CC=/opt/toss/openmpi/4.1/gnu/bin/mpicc
CC -show:
gcc -I/opt/toss/openmpi/4.1/gnu/include -pthread -L/opt/toss/openmpi/4.1/gnu/lib -Wl,-rpath -Wl,/opt/toss/openmpi/4.1/gnu/lib -Wl,--enable-new-dtags -lmpi
CC version:	gcc (GCC) 11.3.0

CXX=/opt/toss/openmpi/4.1/gnu/bin/mpicxx
CXX -show:
g++ -I/opt/toss/openmpi/4.1/gnu/include -pthread -L/opt/toss/openmpi/4.1/gnu/lib -Wl,-rpath -Wl,/opt/toss/openmpi/4.1/gnu/lib -Wl,--enable-new-dtags -lmpi
CXX version:	g++ (GCC) 11.3.0

FC=/opt/toss/openmpi/4.1/gnu/bin/mpif90
FC -show:
gfortran -I/opt/toss/openmpi/4.1/gnu/include -pthread -I/opt/toss/openmpi/4.1/gnu/lib -L/opt/toss/openmpi/4.1/gnu/lib -Wl,-rpath -Wl,/opt/toss/openmpi/4.1/gnu/lib -Wl,--enable-new-dtags -lmpi_usempif08 -lmpi_usempi_ignore_tkr -lmpi_mpifh -lmpi
FC version:	GNU Fortran (GCC) 11.3.0

F77=/opt/toss/openmpi/4.1/gnu/bin/mpif77
F77 -show:
gfortran -I/opt/toss/openmpi/4.1/gnu/include -pthread -I/opt/toss/openmpi/4.1/gnu/lib -L/opt/toss/openmpi/4.1/gnu/lib -Wl,-rpath -Wl,/opt/toss/openmpi/4.1/gnu/lib -Wl,--enable-new-dtags -lmpi_usempif08 -lmpi_usempi_ignore_tkr -lmpi_mpifh -lmpi
F77 version:	GNU Fortran (GCC) 11.3.0

F90=/opt/toss/openmpi/4.1/gnu/bin/mpif90
F90 -show:
gfortran -I/opt/toss/openmpi/4.1/gnu/include -pthread -I/opt/toss/openmpi/4.1/gnu/lib -L/opt/toss/openmpi/4.1/gnu/lib -Wl,-rpath -Wl,/opt/toss/openmpi/4.1/gnu/lib -Wl,--enable-new-dtags -lmpi_usempif08 -lmpi_usempi_ignore_tkr -lmpi_mpifh -lmpi
F90 version:	GNU Fortran (GCC) 11.3.0

OK

##################################################################################################
Python Sanity Checks

Verify `/usr/bin/env python3 --version` (reporting as: Python 3.9.16),
matches versions for: `which python3 && which python`

OK

##################################################################################################
Python Modules (TestHarness, run-ability)

WARNING: python module: `yaml` not available

Warning Python packages may cause miscellaneous runtime issues.

Either install the above packages, or perhaps you have yet to activate your
moose environment: `conda activate moose`.
```

**Reply by GiudGiud** [reply]
did you try to follow the recommendations? e.g. passing --disable-dependency-tracking to the configure option?
this can be done by modifying scripts/configure_libmesh.sh

**Reply by TheGreatCid** [reply]
Yep, I did both. Same error message

**Reply by TheGreatCid** [reply]
Oh wait. I did it with the MOOSE configure script. Oops.

**Reply by TheGreatCid** [reply]
```
Configure complete, now type 'make' and then 'make install'.

---------------------------------------------
--------- Done Configuring libMesh ----------
---------------------------------------------
Running make -j 15...
Makefile:7128: *** missing separator.  Stop.
```

now I get this error

Here is the log file. 
[config.log](https://github.com/user-attachments/files/16875285/config.log)

**Reply by GiudGiud** [reply]
can find the makefile and paste the section around line 7128. see what happened

**Reply by TheGreatCid** [reply]
```
AUTOMAKE = ${SHELL} '/ascldap/users/detorre/raccoon/moose/libmesh/build-aux/missing' automake-1.16
AWK = gawk
AZTECOO_INCLUDES =
AZTECOO_LIBS =
AZTECOO_MAKEFILE_EXPORT =
BOOST_CPPFLAGS = -I/usr/include/
/usr/include/boost1.78/
BOOST_LDFLAGS = -L/usr/lib64
BUILD_ARCH = x86_64-pc-linux-gnu
BUILD_DEVSTATUS = Development Build
BUILD_HOST = mz497
BUILD_USER = detorre
BUILD_VERSION = 3855e00043601e0a45d83ed8b639991c2b288553
BUNZIP2 = bunzip2
BZIP2 = bzip2
CAPNPROTO_INCLUDE =
```

This is from the makefile in `libmesh/build`
`/usr/include/boost1.78/` is line 7128

**Reply by TheGreatCid** [reply]
`git status` does not show any changes in the libmesh directory

**Reply by GiudGiud** [reply]
did you set BOOST_CPPFLAGS?
this new line in the variable is weird!

**Reply by TheGreatCid** [reply]
I didn't 
I didn't touch anything in the build scripts than what was suggested here

**Reply by TheGreatCid** [reply]
maybe I need to do a fresh install?

**Reply by GiudGiud** [reply]
If you set it, without a line break, does it get respected?

**Reply by TheGreatCid** [reply]
It does not. Maybe its a different MakeFile?

**Reply by TheGreatCid** [reply]
Actually, after I run the update libmesh script, the Makefile reverts to its previous state with the line break. I tried this twice to confirm.

**Reply by GiudGiud** [reply]
ah yes you need to skip the submodule update to avoid refreshing the entire folder
with --skip-submodule-update

**Reply by TheGreatCid** [reply]
Sorry for the delay. I had some issues with the hpc that I needed to sort out. 
Where do I insert `skip-submodule-update`? I put it into the configure libmesh file but it seems to not a be a recognized option

**Reply by GiudGiud** [reply]
It goes after update_and_rebuild_libmesh.sh, to prevent it from updating and removing your changes to libmesh

**Reply by TheGreatCid** [reply]
`./update_and_rebuild_libmesh.sh --skip-submodule-update` 
Is this correct? The makefile is still being reset after running.

**Reply by GiudGiud** [reply]
the makefile must be generated automatically.

**Reply by TheGreatCid** [reply]
I did try a fresh install as well, but I am running into the same issue

**Comment by GiudGiud** [comment]
I think it might be time to get help from @roystgnr on this

**Comment by roystgnr** [comment]
Could we see the `config.log` from TIMPI?  It'll be in `contrib/timpi/config.log`

Could you try just running `./bootstrap` in the `contrib/timpi` directory, and let us know the output?

That newline is weird, but it's also a warning and I think it might be a red herring.

**Reply by TheGreatCid** [reply]
![image](https://github.com/user-attachments/assets/0145f8f5-2878-421f-8a8e-74e74a9612d1)
I am not able to see a config file at `contrib/timpi/config.log` 

```
[detorre@manzano-login5 timpi]$ ./bootstrap
Bootstrapping using autoreconf -v -f -i ...
autoreconf: Entering directory `.'
autoreconf: configure.ac: not using Gettext
autoreconf: running: aclocal --force -I m4 -I m4/autoconf-submodule
autoreconf: configure.ac: tracing
autoreconf: running: libtoolize --copy --force
libtoolize: putting auxiliary files in AC_CONFIG_AUX_DIR, 'build-aux'.
libtoolize: copying file 'build-aux/ltmain.sh'
libtoolize: putting macros in AC_CONFIG_MACRO_DIRS, 'm4'.
libtoolize: copying file 'm4/libtool.m4'
libtoolize: copying file 'm4/ltoptions.m4'
libtoolize: copying file 'm4/ltsugar.m4'
libtoolize: copying file 'm4/ltversion.m4'
libtoolize: copying file 'm4/lt~obsolete.m4'
autoreconf: running: /usr/bin/autoconf --force
autoreconf: running: /usr/bin/autoheader --force
autoreconf: running: automake --add-missing --copy --force-missing
configure.ac:86: installing 'build-aux/compile'
configure.ac:14: installing 'build-aux/missing'
src/Makefile.am: installing 'build-aux/depcomp'
autoreconf: Leaving directory `.'
```

Here is the bootstrap output

**Reply by roystgnr** [reply]
`contrib/timpi/config.log` in the build directory (the one where libMesh `config.log` is), not the source directory, is what I'm hoping to see.

Good to see that the bootstrap looks clean on your system at least.

**Reply by TheGreatCid** [reply]
[config.log](https://github.com/user-attachments/files/16965925/config.log)
Here is the config file

**Reply by roystgnr** [reply]
I'm more confused than ever.  The TIMPI `config.log` here ends with `configure: exit 0` (i.e. success), after getting past `executing depfiles commands` just fine.  The main console output above hits `executing depfiles commands` for TIMPI and seems to think (and the main `config.log` agrees) that TIMPI `configure` exited with failure at that point.  These are definitely all logs from the same single `update_and_rebuild_libmesh.sh` run?

**Reply by TheGreatCid** [reply]
I just ran it again
Here is from `libmesh/build`
[config.log](https://github.com/user-attachments/files/16966611/config.log)

Here is from `libmesh/build/contrib/timpi`
[config.log](https://github.com/user-attachments/files/16966614/config.log)

**Reply by TheGreatCid** [reply]
though this is with the `--disable-dependency-tracking` option enabled

**Reply by TheGreatCid** [reply]
Without that option here is the libmesh config
[config.log](https://github.com/user-attachments/files/16966705/config.log)
and the timpi config
[config.log](https://github.com/user-attachments/files/16966707/config.log)

**Reply by roystgnr** [reply]
Okay, that second pair makes more sense; there we're getting the `exit 1` from TIMPI `configure` ... and the `*** missing separator` failure higher up actually does look like it could be from that newline we were warned about, screwing up the whitespace that a Makefile is sensitive to.

That first pair of logs ... seems to be completing fine, with the "Done Configuring libMesh" message and an `exit 0` from libMesh configure ... and then we hit `*** missing separator` as soon as we try to build, because we didn't actually fix any broken Makefiles, we just delayed the point at which they get used and the brokenness gets discovered.  Looking at the logs here I'm not surprised; that newline seems to be corrupting a half dozen different `FLAGS` variables.

... but it looks like it's coming *from* `BOOST_CPPFLAGS`, not the other way around, in that case.  And `BOOST_CPPFLAGS` is coming out of the `AX_BOOST_BASE` macro ... which is ugly complicated.

What does `find /usr/ -name 'shared_ptr.hpp'` see on your system?

For your own sake, you might be able to get around this bug by passing something like `--with-boost=/usr/include/boost1.78` instead of `--disable-dependency-tracking`, but even if that works I hope you'll still be able to help me figure out what's causing this.  Our last change to `ax_boost_base.m4` was over a decade ago and we've never seen anything like this before.

**Reply by TheGreatCid** [reply]
For `find /usr/ -name 'shared_ptr.hpp'
```
[detorre@mz77 ~]$ find /usr/ -name 'shared_ptr.hpp'
/usr/include/boost1.78/boost/exception/detail/shared_ptr.hpp
/usr/include/boost1.78/boost/thread/csbl/memory/shared_ptr.hpp
/usr/include/boost1.78/boost/shared_ptr.hpp
/usr/include/boost1.78/boost/interprocess/smart_ptr/shared_ptr.hpp
/usr/include/boost1.78/boost/serialization/shared_ptr.hpp
/usr/include/boost1.78/boost/smart_ptr/shared_ptr.hpp
/usr/include/boost/exception/detail/shared_ptr.hpp
/usr/include/boost/thread/csbl/memory/shared_ptr.hpp
/usr/include/boost/shared_ptr.hpp
/usr/include/boost/interprocess/smart_ptr/shared_ptr.hpp
/usr/include/boost/serialization/shared_ptr.hpp
/usr/include/boost/smart_ptr/shared_ptr.hpp
```

Your suggestion to specify the boost directory did actually work!

**Reply by roystgnr** [reply]
Well, hopefully it worked.  Seeing both `/usr/include/boost1.78/boost` and `/usr/include/boost` on the same system is scary.  Hopefully the latter is just a symbolic link to the former?

If so, then you should be good to go.  I'll see if I can make such a symlink on my laptop to replicate the problem and fix our boost m4 script for that case.

If they're not just the same directory via links in the filesystem, but they are the same version or ABI-compatible versions, then you should also be good to go, at least until your package manager decides to upgrade one but not the other.

If they're ever ABI-incompatible versions ... something could break, in theory?  Dependency A detects /usr/include/boost1.78 and creates a struct according to one definition, then passes a pointer it to dependency B which detects /usr/include/boost and looks for member offsets according to a different definition, and then you segfault if you're lucky or you get weird corruption if you're not.  I've never had this happen with boost before, but I have seen it with other libraries, and it's even less fun to diagnose than this configure issue was.  If that `/usr/include/boost` isn't just a symlink then I'd bring it up with your HPC guys; you pretty much never want to have multiple versions of software in system directories where they're likely to be autodetected.

**Reply by TheGreatCid** [reply]
It does appear that `/usr/include/boost1.78/boost`  and `/usr/include/boost` are indeed different directories. However, I'm unsure how to check the compatibility of the versions.

I do see what you are saying. They both certainly should not exist.
Looks like this whole debacle came from an issue specific to my HPC rather than anything with MOOSE. 
I really appreciate the help in narrowing this down!

**Reply by roystgnr** [reply]
There's a `version.hpp` you could check.

**Reply by TheGreatCid** [reply]
Sorry for the last response. 
Looking at the headers, one appears to be 1.66 and the other 1.78. So definitely a conflict there

**Reply by roystgnr** [reply]
Oh, yikes.  So if any other dependency puts a `-I/usr/include` in your compiler flags (which is *really* common) then that potential conflict can become an actual conflict.

You might try `--with-boost=/usr` and see if that brings in the `/usr/include/boost`, since it's much less likely that another dependency will be demanding `/usr/include/boost1.78/boost`.  If that works, you're safe.

If that doesn't work?  I'd keep going with `--with-boost=/usr/include/boost1.78`, but be very watchful for either weird compiler errors or weird crashes.
